skada/_reweight.py

```diff
@@ -669,3 +674,223 @@ def KLIEP(
         ),
         base_estimator,
     )
+
+
+class MMDTarSReweightAdapter(BaseAdapter):
+    """MMDTarSReweight adapter.
+
+    The idea of MMDTarSReweight is to find an importance estimate \beta(y) such that
+    the Maximum Mean Discrepancy (MMD) divergence between the source input density
+    p_source(x) to its estimate p_target(x) is minimized under the assumption
+    of equal conditional distributions p(x|y) for both source and target domains.
+
+    See Section 3 of [4]_ for details.
+
+    Parameters
+    ----------
+    gamma : float or array like
+        Parameters for the kernels.
+    reg : float, default=1e-10
+        Regularization parameter for the labels kernel matrix.
+    tol : float, default=1e-6
+        Tolerance for the stopping criterion in the optimization.
+    max_iter : int, default=1000
+        Number of maximum iteration before stopping the optimization.
+
+    Attributes
+    ----------
+    `X_source_` : array-like, shape (n_samples, n_features)
+        The source data used for fitting.
+    `y_source_` : array-like, shape (n_samples,)
+        The source labels used for fitting.
+
+    References
+    ----------
+    .. [4] Kun Zhang et. al. Domain Adaptation under Target and Conditional Shift
+           In ICML, 2013.
+    """
+
+    def __init__(self, gamma, reg=1e-10, tol=1e-6, max_iter=1000):
+        super().__init__()
+        self.gamma = gamma
+        self.reg = reg
+        self.tol = tol
+        self.max_iter = max_iter
+
+    def _weights_optimization(self, X_source, X_target, y_source):
+        """Weight optimization"""
+        m, n = X_source.shape[0], X_target.shape[0]
+
+        # check y is discrete or continuous
+        discrete = _find_y_type(y_source) == "classification"
+
+        # compute A
+        L = pairwise_kernels(y_source.reshape(-1, 1), metric="rbf", gamma=self.gamma)
+        K = pairwise_kernels(X_source, metric="rbf", gamma=self.gamma)
+        omega = L @ np.linalg.inv(L + self.reg * np.eye(m))
+        A = omega @ K @ omega.T
+
+        # compute R
+        if discrete:
+            classes = np.unique(y_source)
+            R = np.zeros((X_target.shape[0], len(classes)))
+            for i, c in enumerate(classes):
+                R[:, i] = (y_source == c).astype(int)
+        else:
+            R = L @ np.linalg.inv(L + self.reg * np.eye(m))
+
+        # compute M
+        K_cross = pairwise_kernels(X_target, X_source, metric="rbf", gamma=self.gamma)
+        M = np.ones((1, n)) @ K_cross @ omega.T
+
+        # solve the optimization problem
+        # min_alpha 0.5 * alpha^T P alpha - q^T alpha
+        # s.t. 0 <= R alpha <= B_beta
+        #      m (1 - eps) <= 1^T R alpha <= m (1 + eps)
+        P = R.T @ A @ R
```

Comment:
sems like a generic QP solver will be usefull elsewhere. Do we want to code a function that takes (we can put that in utils) standard QP as input?